src/west/cmd/build.py

```diff
@@ -0,0 +1,255 @@
+        '''Final build directory used to run the build, or None on error.'''
+
+        self.created_build_dir = False
+        '''True if the build directory was created; False otherwise.'''
+
+        self.ran_cmake = False
+        '''True if CMake was successfully run to generate a build system;
+        False otherwise.'''
+
+        self.cmake_args = None
+        '''If CMake was run, arguments used to run it; None otherwise.'''
+
+        self.cmake_cache = None
+        '''Final parsed CMake cache for the build, or None on error.'''
+
+    def do_add_parser(self, parser_adder):
+        parser = parser_adder.add_parser(
+            self.name,
+            formatter_class=argparse.RawDescriptionHelpFormatter,
+            description=self.description)
+
+        parser.add_argument('-b', '--board',
+                            help='''board to build for (must be given for the
+                            first build, can be omitted later)''')
+        parser.add_argument('-s', '--source-dir',
+                            help='''explicitly sets the source directory;
+                            if not given, infer it from directory context''')
+        parser.add_argument('-d', '--build-dir',
+                            help='''explicitly sets the build directory;
+                            if not given, infer it from directory context''')
+        parser.add_argument('-t', '--target',
+                            help='''override the build system target (e.g.
+                            'clean', 'pristine', etc.)''')
+        parser.add_argument('-c', '--cmake', action='store_true',
+                            help='force CMake to run')
+        parser.add_argument('-f', '--force', action='store_true',
+                            help='ignore any errors and try to build anyway')
+        parser.add_argument('cmake_opts', nargs='*', metavar='cmake_opt',
+                            help='extra option to pass to CMake; implies -c')
+
+        return parser
+
+    def do_run(self, args, ignored):
+        self.args = args        # Avoid having to pass them around
+        log.dbg('args:', args, level=log.VERBOSE_EXTREME)
+        self._sanity_precheck()
+        self._setup_build_dir()
+        if is_zephyr_build(self.build_dir):
+            self._update_cache()
+        else:
+            self.args.cmake = True
+        self._setup_source_dir()
+        self._sanity_check()
+
+        log.inf('source directory: {}'.format(self.source_dir))
+        log.inf('build directory: {}{}'.
+                format(self.build_dir,
+                       (' (created)' if self.created_build_dir
+                        else '')))
+        # FIXME needs some extra logic in here
+        if self.cmake_cache:
+            board = self.cmake_cache.get('CACHED_BOARD')
+        else:
+            board = 'UNKNOWN'   # shouldn't happen
+        log.inf('BOARD:', board)
+
+        self._run_cmake(self.args.cmake_opts)
+        self._sanity_check()
```

Comment:
Just to check, is it intentional that `_sanity_check()` is called twice in this function?